src/workerd/jsg/modules-new.c++

```diff
@@ -0,0 +1,1401 @@
+// Copyright (c) 2017-2022 Cloudflare, Inc.
+// Licensed under the Apache 2.0 license found in the LICENSE file or at:
+//     https://opensource.org/licenses/Apache-2.0
+
+#include "modules-new.h"
+#include <workerd/jsg/jsg.h>
+#include <workerd/jsg/function.h>
+#include <workerd/jsg/util.h>
+#include <kj/mutex.h>
+#include <kj/table.h>
+
+namespace workerd::jsg::modules {
+
+namespace {
+kj::Maybe<Module&> checkModule(const ResolveContext& context, Module& module) {
+  if (!module.evaluateContext(context)) {
+    return kj::none;
+  }
+  return module;
+};
+
+bool ensureInstantiated(
+    Lock& js,
+    v8::Local<v8::Module> module,
+    const CompilationObserver& observer,
+    const Module* self) {
+  if (module->GetStatus() == v8::Module::kUninstantiated) {
+    bool result;
+    if (!self->instantiate(js, module, observer).To(&result)) {
+      return false;
+    }
+    if (!result) {
+      js.v8Isolate->ThrowError(
+          js.str(kj::str("Failed to instantiate module: ", self->specifier())));
+      return false;
+    }
+  }
+  return true;
+}
+
+ModuleBundle::Type toModuleBuilderType(ModuleBundle::BuiltinBuilder::Type type) {
+  switch (type) {
+    case ModuleBundle::BuiltinBuilder::Type::BUILTIN:
+      return ModuleBundle::Type::BUILTIN;
+    case ModuleBundle::BuiltinBuilder::Type::BUILTIN_ONLY:
+      return ModuleBundle::Type::BUILTIN_ONLY;
+  }
+  KJ_UNREACHABLE;
+}
+
+class EsModule final : public Module {
+public:
+  explicit EsModule(Url specifier, Type type, Flags flags, kj::Array<const char> source)
+      : Module(kj::mv(specifier), type, flags | Flags::ESM),
+        source(kj::mv(source)),
+        ptr(this->source),
+        externed(false),
+        cachedData(nullptr) {
```

Comment:
Nit: `cachedData` is set to nullptr in one constructor but not the other one – with it also being set to nullptr in its definition I guess doing so here is not needed.